rclcpp/include/rclcpp/copy_all_parameters.hpp

```diff
@@ -0,0 +1,51 @@
+// Copyright 2023 Open Navigation LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef RCLCPP__COPY_ALL_PARAMETERS_HPP_
+#define RCLCPP__COPY_ALL_PARAMETERS_HPP_
+
+#include <vector>
+#include <string>
+#include "rcl_interfaces/srv/list_parameters.hpp"
+#include "rclcpp/parameter.hpp"
+
+namespace rclcpp
+{
+
+/// Copy all parameters from one source node to another destination node.
+/**
+ * \param source Node to copy parameters from
+ * \param destination Node to copy parameters to
+ * \param override Default false. Whether to override destination parameters
+ * if both the source and destination contain the same parameter.
+ */
+template<typename NodeT1, typename NodeT2>
+void
+copy_all_parameters(const NodeT1 & source, const NodeT2 & destination, const bool override = false)
```

Comment:
can we replace the template with a `rclcpp::node_interfaces::NodeParametersInterface`?

Reply:
`override` is a C++ keyword; I would prefer using a differently named variable

Reply:
We could, though I much prefer the templating since it doesn't force the user to include nasty code into their application like `rclcpp::copy_all_parameters(node1->get_node_parameters_interface(), node2->get_node_parameters_interface())` which I think is a far less good user experience. 

I can change override if you like, what would you suggest?

Reply:
> I can change override if you like, what would you suggest?

IMO `override_params` is already better (I'm fine with everything that isn't just `override`).

> I much prefer the templating since it doesn't force the user to include nasty code into their application

As far as I know, passing node interfaces is the recommended approach, as it better scales to all the different "node-like" classes users may want to implement. Ultimately that's what you care about, i.e. that the object passed as argument has a parameter interface.
This also makes it a lot clearer how users should use the function: at a first read I would expect that `const NodeT1 & source` would expect a `rclcpp::Node`, but it actually expect some sort of pointer (e.g. `shared_ptr<rclcpp::Node>`) as it uses the arrow operator.

Reply:
This isn't in a cpp file that would require explicit instantiation of the types, this would still work with any node-like object. 

If the naming would help changing to `source_node_ptr` or something, that could be done as well! Let me know, I just don't want to piecemeal update the PR 5x with minor changes that don't resolve your concerns. (1) node names and (2) override rename so far

Reply:
This would work with the node-like classes that you mentioned (`rclcpp::LifecycleNode`, `nav2_utils::LifecycleNode`, and `rclcpp::Node`), but not with any other custom node-like class that, although it contains a node parameters interface (and thus it should be able to use this function) does not expose the methods used here (`get_parameters`, `set_parameters`, etc.

As a dummy example, I may want to use it with this custom node
```
class MyNode
{
public:
    MyNode();
    rclcpp::node_interfaces::NodeParametersInterface::SharedPtr node_parameters_;
}
```

If directly using parameters interface is not a viable solution, then we should at least use the new `NodeInterface` class (https://github.com/ros2/rclcpp/pull/2041).
This is compatible with both node-like classes and directly passing interfaces.

Reply:
Mhm, this is a good point -- though perhaps a dumb question, couldn't you just literally send it that `node_parameters_`? That is a node like object by itself with those exact APIs defined. I think 

`rclcpp::copy_all_parameters(MyNode1.node_parameters_, MyNode2.node_parameters_)` would work all the same?

Reply:
I generally resist having to make users have to `get_XYZ_base_interface()` in their application code. Its so dirty looking and really mucks things up. I really don't believe that the average ROS user should even need to know about the internal structure of the node enough to be aware of those interfaces to write applications. I know that's not the case right now, today, but I figure I gotta start somewhere :smile:

Reply:
With the current implementation of `copy_all_parameters` it's not possible to directly pass a `rclcpp::node_interfaces::NodeParametersInterface` because it doesn't have a method named `set_parameter` (it has `set_parameterS` though).
Changing the implementation of the function (to use `set_parameters` instead of `set_parameter` would IMO not be the best change, as you are relying on node interfaces and nodes-like objects to have the same exact APIs (which is not enforced anywhere).

I think that using the `NodeInterfaces` class introduced in the PR above is likely the best of both worlds.
It will still allow users to directly pass a `rclcpp::Node` or similar object or to pass directly the node interfaces if a node is not available.

Reply:
Are you suggesting something like:

```
template<typename NodeT1, typename NodeT2>
void
copy_all_parameters(const NodeT1 & source, const NodeT2 & destination, const bool override = false)
{
  using Parameters = std::vector<rclcpp::Parameter>;
  auto source_params = NodeInterfaces<NodeParametersInterface>(source).get_node_parameters_interface();
  auto dest_params = NodeInterfaces<NodeParametersInterface>(destination).get_node_parameters_interface();

  std::vector<std::string> param_names = source_params->list_parameters({}, 0).names;
  Parameters params = source_params->get_parameters(param_names);
  for (Parameters::const_iterator iter = params.begin(); iter != params.end(); ++iter) {
    if (!dest_params->has_parameter(iter->get_name())) {
      dest_params->declare_parameter(iter->get_name(), iter->get_parameter_value());
    } else if (override) {
      dest_params->set_parameter(*iter);
    }
  }
}
```

I can see the potential issue with a simple `NodeT` option if I'm just calling direct APIs from it if some non-standard-node doesn't implement the same API as all current nodes. I can get behind that idea; we shouldn't assume _too_ too much about the nature of the Node where reasonable accommodation can be made. 

Separate semi-rant: I went through that PR with `NodeInterfaces` and can understand the value in providing little to no assumptions about the Node-like object - though it feels **dramatically** over-engineered. This is the kind of obscure complexity creep that leaves a sour taste in my mouth when I look over some of the ROS 2 core codebases. I don't think this level of trying to accommodate all possible strange situations via these complex solutions is merited and really draws back on potential contributions to support extremely obscure needs. This solution **still** assumes that every node object implements the `get_node_XYZ_interface()`!  And the XYZ is still hardcoded in the software's interface definitions, so its not like this actually makes it more modular! 

```
    get_from_node_like(NodeT & node_like) \
    { \
      return node_like.get_node_ ## NodeInterfaceName ## _interface(); \
    } \
```

If we place this really minimal and reasonable constraints on all Node-like objects you'd like to be able to interact with the ROS Node-centric API that we implement the `get_node_XYZ_interface()` methods, then we can trivially do

 ```
template<typename NodeT1, typename NodeT2>
void
copy_all_parameters(const NodeT1 & source, const NodeT2 & destination, const bool override = false)
{
  using Parameters = std::vector<rclcpp::Parameter>;
  auto source_params = source->get_node_parameters_interface();
  auto dest_params = destination->get_node_parameters_interface();

  std::vector<std::string> param_names = source_params->list_parameters({}, 0).names;
  Parameters params = source_params->get_parameters(param_names);
  for (Parameters::const_iterator iter = params.begin(); iter != params.end(); ++iter) {
    if (!dest_params->has_parameter(iter->get_name())) {
      dest_params->declare_parameter(iter->get_name(), iter->get_parameter_value());
    } else if (override) {
      dest_params->set_parameter(*iter);
    }
  }
}
```

which is no different really in terms of the core assumptions made without all this casting nonsense that makes the code far less clear in intent. All we're really doing is moving the location of the assumed name and adding in layers and layers of nearly pointless templating! 

This seems like insanity for insanity's sake!

Reply:
For the scope of this PR - I suppose we just need to come to the agreement about whether to use the NodeInterfaces object or I can just access the interface `getter` to meet the generalization need you mention and I agree with

Reply:
I mean we added `NodeInterfaces` for this reason, so it would make sense to use it imo. But we added it to make things easier/better, so if the benefit isn't there for this use case then maybe we should discuss it more.

Mostly I'd say use it just so we have consistency (or at least we're moving in the direction of consistency we want).

Reply:
I forgot about it when I reviewed this originally.

Reply:
> Are you suggesting something like:

Sorry still catching up on the convo, but I think the suggestion is:

```suggestion
copy_all_parameters(NodeInterfaces<NodeParametersInterface> source, NodeInterfaces<NodeParametersInterface> destination, const bool override = false)
```

And everything else is the same?

Well except you can also do things like `auto npi = source.get<NodeParametersInterface>()` instead of `auto npi = source.get_node_parameters_interface()`, which I think is nicer.

Have a look at the test cases for more examples:

https://github.com/ros2/rclcpp/blob/ea31f94eb453acf3f34b37b85e38284abc514d57/rclcpp/test/rclcpp/node_interfaces/test_node_interfaces.cpp#L48-L190

Reply:
Sorry, I guess it's more complex than that, since you're actually doing more than using individual node interfaces, instead you're using methods which may only exist on the `Node` type itself (and some others like the `LifecycleNode` types).

Ideally, these methods on Node would instead be free functions, sort of like how we have free function versions of `create_publisher()`, which is what `Node::create_publisher()` calls. But that might not be the case.

However, if you can limit yourself to what's available on the `NodeParametersInterface` then it will be strictly more portable/flexible. I'd even be in favor of factoring out any syntactic sugar you need from methods on the `Node` class into free functions if that's not already the case, e.g. for things like `declare_parameter`, or `list_parameter`, etc. And it is all syntactic sugar, since they all end up going through `NodeParametersInterface` eventually anyway.

Even more ideally, we'd have only free functions and C++ would implement the "Unified Call Syntax" proposal (https://brevzin.github.io/c++/2019/04/13/ufcs-history/ just one resource, lots on the internet about it), but I won't hold my breathe :(

Reply:
>  it would make sense to use it imo.

@wjwwood its been in the project for 9 months and I just grepped the codebase and it has 0 uses outside of its implementation and unit tests. I would resist making this tiny utility the thing that brings it into production. A larger effort should be made to make other things portable before we start nit-picking new contributions to do so (eg have consistency, like changing tests to use current codebase styling).

Also: 

> copy_all_parameters(NodeInterfaces<NodeParametersInterface> source, NodeInterfaces<NodeParametersInterface> destination, const bool override = false)

looks nasty as heck, unless it automatically casts from the standard node types? Some remark in the original PR makes me think it may (?). Either way, that's an icky looking API that would be quite confusing to try to understand as an external user of the API. That's creating some real insider baseball. 

> you're using methods which may only exist on the Node type itself (

I'd update that one API to use parameter**S** to make it all use the interface API methods that are available. That should be possible, I just didn't update that one line for only illustration purposes :smile:

Reply:
@wjwwood @alsora thoughts?

Reply:
> unless it automatically casts from the standard node types

Yes, it casts automatically from node types (see example in the unit-tests https://github.com/ros2/rclcpp/blob/rolling/rclcpp/test/rclcpp/node_interfaces/test_node_interfaces.cpp#L59-L67)

> I'd update that one API to use parameterS

IMO that's not the correct approach as nothing requires that APIs in the node and in the underlying interface match.

 ----- 

This definitely evolved into a much larger discussion than what the PR was trying to address.
I would be ok to go with what is IMO a short-term approach (either using the parameter**S** API or using templates) only if we (ROS maintainers) prioritize fixing any lingering issue with the `NodeInterfaces` class.

Having APIs that take nodes as arguments is something that makes it harder to adhere to the best-practices to optimize perfomrance and have really modular ROS 2 applications (examples are passing nodes around or creating nodes within nodes).

If we want ROS to succeed in a variety of use-cases, we must make sure that it's both easy to use for beginners as well as possible to tune for advanced users.
As it stands today, there's still lot of work needed in both directions.

Reply:
> and in the underlying interface match.

I think the entire system would break if we didn't assume that the ROS node abstract interfaces were implemented. The `NodeInterfaces` method assumes the exact same things as I am as I pointed out above. The `NodeInterfaces` direction equally relies on the parameters interface existing and gettable via `get_node_XYZ_interface()`. They're functionally identical but without the seemingly nutty levels of indirection. If we look at it from a requirements direction they both assume the following, no more, no less (if I use the `node_parameters->set_parameters` API:
- That there is some method called `get_node_parameter_interfaces()` that returns a ROS parameter interface
- That the interface contains the same get / set implementations (with **s**) 

If we can't agree on that, I don't see how we can assume anything about a node to accomplish anything in rclcpp. That is already the lowest common denominator. I'm not sure we can get any lower.

Reply:
To be clear, I'm suggesting:
- I'll get the param interface via `get_node_parameter_interface()`  for each node
- Use the methods that the node parameters interface provides to me **only**, not `node->api()`, but `node->get_node_parameter_interface()->api()` 

Is that OK? I don't see how that's any different from the `NodeInterfaces` direction other than creating obscurity